dhcpv4/client.go

```diff
@@ -267,31 +286,61 @@ func (c *Client) sendReceive(sendFd, recvFd int, packet *DHCPv4, messageType Mes
 	// Create a goroutine to perform the blocking send, and time it out after
 	// a certain amount of time.
 	var (
-		destination [4]byte
+		destination [net.IPv4len]byte
 		response    *DHCPv4
 	)
 	copy(destination[:], raddr.IP.To4())
 	remoteAddr := unix.SockaddrInet4{Port: laddr.Port, Addr: destination}
 	recvErrors := make(chan error, 1)
 	go func(errs chan<- error) {
-		conn, innerErr := net.FileConn(os.NewFile(uintptr(recvFd), ""))
-		if innerErr != nil {
+		// set read timeout
+		timeout := unix.NsecToTimeval(c.ReadTimeout.Nanoseconds())
+		if innerErr := unix.SetsockoptTimeval(recvFd, unix.SOL_SOCKET, unix.SO_RCVTIMEO, &timeout); innerErr != nil {
 			errs <- innerErr
 			return
 		}
-		defer conn.Close()
-		conn.SetReadDeadline(time.Now().Add(c.ReadTimeout))
-
 		for {
 			buf := make([]byte, MaxUDPReceivedPacketSize)
-			n, _, _, _, innerErr := conn.(*net.UDPConn).ReadMsgUDP(buf, []byte{})
+			n, _, innerErr := unix.Recvfrom(recvFd, buf, 0)
 			if innerErr != nil {
 				errs <- innerErr
 				return
 			}
 
-			response, innerErr = FromBytes(buf[:n])
+			var iph ipv4.Header
+			if err := iph.Parse(buf[:n]); err != nil {
+				// skip non-IP data
+				continue
+			}
+			if iph.Protocol != 17 {
+				// skip non-UDP packets
+				continue
+			}
+			udph := buf[iph.Len:n]
+			// check source and destination ports
+			srcPort := int(binary.BigEndian.Uint16(udph[0:2]))
+			expectedSrcPort := ServerPort
+			if c.RemoteAddr != nil {
+				expectedSrcPort = c.RemoteAddr.(*net.UDPAddr).Port
+			}
+			if srcPort != expectedSrcPort {
+				continue
+			}
+			dstPort := int(binary.BigEndian.Uint16(udph[2:4]))
+			expectedDstPort := ClientPort
+			if c.RemoteAddr != nil {
+				expectedDstPort = c.LocalAddr.(*net.UDPAddr).Port
+			}
+			if dstPort != expectedDstPort {
+				continue
+			}
+			// UDP checksum is not checked
+			pLen := int(binary.BigEndian.Uint16(udph[4:6]))
+			payload := buf[iph.Len+8 : iph.Len+8+pLen]
+
+			response, innerErr = FromBytes(payload)
 			if innerErr != nil {
+				log.Print(payload)
```

Comment:
Do we want to print the payload or is this left from some testing?

Reply:
leftover, thanks!